modules/pam_setquota/pam_setquota.c

```diff
@@ -0,0 +1,238 @@
+	 Copyright © 2010 Shane Tzen <shane@ict.usc.edu>
+	 Copyright © 2012 Sven Hartge <sven@svenhartge.de>
+   Copyright © 2016 Keller Fuchs <kellerfuchs@hashbang.sh>
+
+*/
+
+#include <sys/types.h>
+#include <sys/quota.h>
+#include <linux/quota.h>
+#include <pwd.h>
+#include <syslog.h>
+#include <errno.h>
+#include <mntent.h>
+#include <stdio.h>
+#include <stdbool.h>
+
+#define PAM_SM_SESSION
+
+#include <security/pam_modules.h>
+#include <security/_pam_macros.h>
+#include <security/pam_ext.h>
+#include <security/pam_modutil.h>
+
+#define LOGIN_DEFS "/etc/login.defs"
+
+struct pam_params {
+  uid_t start_uid;
+  uid_t end_uid;
+  char *fs;
+  int overwrite;
+  int debug;
+};
+
+static inline void debug(pam_handle_t *pamh, struct if_dqblk *p) {
+  pam_syslog(pamh, LOG_DEBUG, "bsoftlimit=%llu bhardlimit=%llu"
+                              "isoftlimit=%llu ihardlimit=%llu",
+             p->dqb_bsoftlimit, p->dqb_bhardlimit, p->dqb_isoftlimit,
+             p->dqb_ihardlimit);
+}
+
+static bool _pam_parse_dqblk(int argc, const char **argv, struct if_dqblk *p) {
+  bool bhard = false, bsoft = false, ihard = false, isoft = false;
+
+  for (; argc-- > 0; ++argv) {
+    if (strncmp(*argv, "bhardlimit=", 11) == 0) {
+      p->dqb_bhardlimit = strtol(*argv + 11, NULL, 10);
+      p->dqb_valid |= QIF_BLIMITS;
+      bhard = true;
+    } else if (strncmp(*argv, "bsoftlimit=", 11) == 0) {
+      p->dqb_bsoftlimit = strtol(*argv + 11, NULL, 10);
+      p->dqb_valid |= QIF_BLIMITS;
+      bsoft = true;
+    } else if (strncmp(*argv, "ihardlimit=", 11) == 0) {
+      p->dqb_ihardlimit = strtol(*argv + 11, NULL, 10);
+      p->dqb_valid |= QIF_ILIMITS;
+      ihard = true;
+    } else if (strncmp(*argv, "isoftlimit=", 11) == 0) {
+      p->dqb_isoftlimit = strtol(*argv + 11, NULL, 10);
+      p->dqb_valid |= QIF_ILIMITS;
+      isoft = true;
+    } else if (strncmp(*argv, "btime=", 6) == 0) {
+      p->dqb_btime = strtol(*argv + 6, NULL, 10);
+      p->dqb_valid |= QIF_BTIME;
+    } else if (strncmp(*argv, "itime=", 6) == 0) {
+      p->dqb_itime = strtol(*argv + 6, NULL, 10);
+      p->dqb_valid |= QIF_ITIME;
+    }
+  }
+
+  return !(bhard ^ bsoft) && !(ihard ^ isoft);
+}
+
+static void _pam_parse_params(int argc, const char **argv,
+                              struct pam_params *p) {
+  for (; argc-- > 0; ++argv) {
+    if (strncmp(*argv, "startuid=", 9) == 0)
+      p->start_uid = strtol(*argv + 9, NULL, 10);
+    else if (strncmp(*argv, "enduid=", 7) == 0)
+      p->end_uid = strtol(*argv + 7, NULL, 10);
+    else if (strncmp(*argv, "fs=", 3) == 0)
+      p->fs = (char *)*argv + 3;
+    else if (strncmp(*argv, "overwrite=", 10) == 0)
+      p->overwrite = strtol(*argv + 10, NULL, 10);
+    else if (strncmp(*argv, "debug=", 6) == 0)
+      p->debug = strtol(*argv + 6, NULL, 10);
+  }
+}
+
+PAM_EXTERN int pam_sm_open_session(pam_handle_t *pamh, int flags, int argc,
+                                   const char **argv) {
+  int retval;
+	char *val;
```

Comment:
Indentation seems wrong - tabs/vs spaces?

Reply:
When you once use an editor on a different machine... Fixed in 3960007